### ISSUE TITLE
Multi GPU system - miners not activating all GPUS

### ISSUE THREAD

**Issue by Splitter9980** [aaronsace/MultiPoolMiner#522]
This will obviously seem naive, but I am.

I just built a 6 GPU (NVIDIA 1070) system and I’m finding that a fairly large number of miners do not scale up to the full potential of the system as well as others. Running all algo benchmarks for Miningpoolhub and Zpool I am noticing that sometimes only 1-2 GPU’s will ‘activate’ and the rest remain idle, seemingly at random because I can’t find a root cause.  Certain combinations of miners and algorithms, with no commonality between them.

Comparing the benchmark results to a single 1070 GPU system, it’s not the straightforward x6 I was expecting. Is this a known issue that I just haven’t figure out how to get around yet?

Initially I was running in to page file issues, but now I have page file set to 60GB and I’m only seeing max 70% utilization of that, so that should no longer be the issue. What should I be looking at next?

**Comment by Splitter9980** [comment]
Here is a screenshot of GPU utilization from Afterburner during the benchmark process where I have crudely marked algo start/stop time (red and green vertical lines) and then indicated pass (tick) fail (cross) delay (dash) for if the GPU was working during the testing time. This is a random time period, I did not queue up these algos/miners to get this result.

Interesting to see that sometime a card will start to compute but then fail (?) and return to idle.

![ba5caaa5-c667-4274-a19b-29d66007a87b](https://user-images.githubusercontent.com/33817843/33114233-96d7bbb0-cf21-11e7-9511-13784a3743a1.jpeg)

**Comment by lyolyalya** [comment]
that's the problem with alexis miner.he didn't test it with large number of gpus(3+).there some algos that have that problems - skein,x17,sib,nist5... you need to decrease default intensty of miners or to use "--cuda-shedule spin" flag.but the last is causing cpu to be 100% load...

**Comment by Splitter9980** [comment]
OK, so how do I decrease default intensity for those miners when I assume MultiPoolMiner handles this?

**Comment by lyolyalya** [comment]
i don't use MPM now but anywhere where it possible to cusomize miners start parameters like c=symbol or d=diff...

**Comment by aaronsace** [comment]
There is a section at the top of each ccminer file in the folder called 'miners' to enter advanced customisation parameters for each algorithm.

**Comment by Splitter9980** [comment]
I had a look, but this is getting way beyond my knowledge. Is the default difficulty set somewhere that I could just overwrite, or do I have to go each ccminer file at a time?

I couldn’t decipher where, or how, to enter difficulty either, if there is something I could read to understand that would be great.

**Comment by aaronsace** [comment]
Multi-GPU systems benefit from a good CPU; a lot of people just get the cheapest thinking it doesn't matter at all. The CPU is integral for the whole system. 

Some miners have higher CPU usage than others but would require some experimenting. 

I think some of the miners have a command line option to stager GPU initialization but it would need some research.

**Comment by Splitter9980** [comment]
I may have found it, is it the $Commands section? And use “-i” to set intensity?

How do I know what intensity to set it to tho?

**Comment by Splitter9980** [comment]
Oh, yeah, and that’s me with a cheap celeron... Made a few mistakes on this build that I wouldn’t do again

**Comment by aaronsace** [comment]
Edit each ccminer file at a time. Put a 'space' then the option like this: " -i 20"

**Comment by Splitter9980** [comment]
Thanks! I’ll try and let you know how it goes

**Comment by lyolyalya** [comment]
@Splitter9980 try to decrease it by one until you get all cards working.for my rig it was 27 to 21 on skein algo for example.

**Comment by UselessGuru** [comment]
A suggestion to @aaronsace 
To make troubleshooting / intensity finding easier I added the following line to MPM just right before where it starts the miner bin:

`"$(Split-Path $_.Path -leaf) $($_.Arguments)" | Out-File "$(Split-Path $_.Path)\$($_.Name)_$($($_.Algorithm | ForEach-Object { (($_)+ "_" + $Pools.$_.Name)}) -join "-").cmd"`

Like this you will get a cmd file for each run miner/algo/pool combination in the bin directory.
To find the best intensity I just take the content of the cmd file and run it with different intensities.

**Comment by aaronsace** [comment]
@UselessGuru I think the trouble is that you will have the process ID of command prompt instead of the actual miner so you risk closing command prompt but the miner process is still running.

**Comment by UselessGuru** [comment]
@aaronsace 
???
This the result of the line above:

> ccminer.exe -a neoscrypt -o stratum+tcp://neoscrypt.eu.nicehash.com:3341 -u 3QYQg9JwVSvTTwpAdfaXJe2jQPC8JhG4X8.1080ti -p x -i 22 

I use this output as a starting point for finding the optimal intensity. It still a manual task, but at least I have all the needed parameters prepared already.

> Need to think about it; the whole intensity thing needs simplifying.

Agree, but I cannot thing of an automated way - Anyway I see the whole intensity thing as something for advanced users.

**Comment by aaronsace** [comment]
@UselessGuru 
Oh, I see; so you aren't running the CMD within MultiPoolMiner. That's a good way to experiment with the config.

**Comment by Splitter9980** [comment]
Ok, so going one by one compared to my single GPU stats and getting there slowly. You can monitor the ccminer window also which shows the initial intensity, but that's time consuming.

Watching CPU stats and I'm seeing a lot of 100% utilisation so it's time to get a new CPU, what do you recommend on the LGA1151 chipset that would be sufficient? I'm think i3 7100 but don't want to land short on CPU power again.

**Comment by spider703** [comment]
in the Miners folder open the editor for example CcminerTpruvot.ps1 and change "x17" = "" # X17 on in "x17" = " -i 20" # X17
This miner on this algorithm will work on the intensity of 20

**Comment by fonyo** [comment]
Questions around customising configs for specific cards are now coming up more often, I will draft some general guidance on this and add it to the README. In the future (and for V3's smart GPU recognition and handling), **it might be worth gathering this data and for now adding to the miner files as commented out lines**. 

I remember writing a config generator for sgminer in PHP and running a website around it a few years ago. You could add your address and worker details, pick pools or P2pool, choose user-friendly intensities, dev donations, load balancing etc. For NVIDIA/ccminer this would be very simple to replicate as you only need the intensity data.

**Comment by Splitter9980** [comment]
New PCU in, and 7th GPU. Looking much better, CPU stats under control and GPU utilization is much better. Thanks for the help folks, I will report back at the end of benchmarking, but I like where this software is going!
![802458e0-cc0d-49a6-84ce-897861b93ce1](https://user-images.githubusercontent.com/33817843/33236582-7a08d2e0-d21d-11e7-920f-98f2793ad217.jpeg)

**Comment by lyolyalya** [comment]
what CPU are you using now?

**Comment by Splitter9980** [comment]
i3-7100

**Comment by lyolyalya** [comment]
thanks.

**Comment by Coin-Digger** [comment]
@Splitter9980 
so the solution for a multi GPU rig is a high end CPU to avoid activation gap between GPUs, by the way, have you tried to change any parameter setting still works with a low end CPU like Celeron?  Now I am also going to build a 8xGPU with Nvidia 1070 VGA cards, just want to know from you the feasibility with such low end CPU only with changing some parameter settings.

**Comment by esenocak** [comment]
i'm using 13x1080ti with i3 7100 but delayed activation of gpus is an issue with alexis miner (and it's variants), tpruvot and klaust have no issue so far and switching good

**Comment by UselessGuru** [comment]
> i'm using 13x1080ti with i3 7100 but delayed activation of gpus is an issue with alexis miner (and it's variants), tpruvot and klaust have no issue so far and switching good

Note: Not all miners can handle 13 cards successfully.
Your CPU seem very weak for this kind of setup. I am sure you would ses faster hash rates with a faster processor.

**Comment by Braintelligence** [comment]
@UselessGuru I'm running 13 GPUs on a Pentium G4560 on three different rigs. No problem whatsoever with CPU load.

**Comment by esenocak** [comment]
i also agree that that is because of the intensity, alexis (and variants) and kalust miners uses intensity of >25 for most algos, but on the other hand tpruvot seems constantly using i=20, that may be the one reason, whatever your cpu power (oc at least should be 4 threads cpu) is the second cause maybe but i do not think so.

**Comment by UselessGuru** [comment]
When using a lots of GPUs you also allocate need to allocate huge amounts of virtual memory, especially when using high intensities. Not enough RAM (incl. swap space) will cause miner crashes.
As a rule of thumb I use 2x the amount of memory installed on all GPUs (eg. 10x 1080ti = 10 x 11GB x2 = 220GB)

**Comment by esenocak** [comment]
@UselessGuru then we need at least 256 or 512 ssd's for higher amount of gpus, how is your switching performance with that settings? i tried to increase the virtual mem up to 96gb's but nothing changed.

**Comment by kodeballer** [comment]
just tweek the intensity, i'm currently doing this and seeing positive results, or a cut off point where i just won't bother w/ that algorithm.

**Comment by lyolyalya** [comment]
to activate all gpus on alexis miner use --cuda-schedule spin.
but be prepared to 100% cpu load with it.

**Comment by UselessGuru** [comment]
> we need at least 256 or 512 ssd's for higher amount of gpus

For exact settings you need to ask the miner developer - this is out of scope of MPM. I just made a recommendation based on information I gather over time.
